From 54ce74b5c7e4fd8750b443d8628ea047a1095473 Mon Sep 17 00:00:00 2001
From: Unknown <unknown@example.com>
Date: Thu, 28 Mar 2024 00:37:14 +0000
Subject: [PATCH] This hack stops kernel oops at alsa cards caps query. By
 https://github.com/torvalds/linux/commit/446b31e894935ebbcf84302061a4e0e2efb2368f
 h616 audio patches were updated. After this - h616 audio works ok but we
 started to see kernel oops at alsa cards caps query.

This is workarround patch:
-reverts 446b31e894935ebbcf84302061a4e0e2efb2368f (hunks 1,2)
-reverts h616 audio patches chnages required by 446b31e894935ebbcf84302061a4e0e2efb2368f (hunks 3,4)
---
 include/sound/soc-dai.h             | 13 +++++++++++
 sound/soc/soc-core.c                | 25 +++++++++++++++++++++
 sound/soc/sunxi/sun50iw9-codec.c    |  5 +----
 sound/soc/sunxi_v2/snd_sunxi_ahub.c | 34 ++++++++++++++---------------
 4 files changed, 56 insertions(+), 21 deletions(-)

diff --git a/include/sound/soc-dai.h b/include/sound/soc-dai.h
index adcd8719d343..872dc5cee8fe 100644
--- a/include/sound/soc-dai.h
+++ b/include/sound/soc-dai.h
@@ -415,6 +415,15 @@ struct snd_soc_dai_driver {
 	struct snd_soc_dobj dobj;
 	struct of_phandle_args *dai_args;
 
+	/* DAI driver callbacks */
+	int (*probe)(struct snd_soc_dai *dai);
+	int (*remove)(struct snd_soc_dai *dai);
+	/* compress dai */
+	int (*compress_new)(struct snd_soc_pcm_runtime *rtd, int num);
+	/* Optional Callback used at pcm creation*/
+	int (*pcm_new)(struct snd_soc_pcm_runtime *rtd,
+		       struct snd_soc_dai *dai);
+
 	/* ops */
 	const struct snd_soc_dai_ops *ops;
 	const struct snd_soc_cdai_ops *cops;
@@ -425,6 +434,10 @@ struct snd_soc_dai_driver {
 	unsigned int symmetric_rate:1;
 	unsigned int symmetric_channels:1;
 	unsigned int symmetric_sample_bits:1;
+
+	/* probe ordering - for components with runtime dependencies */
+	int probe_order;
+	int remove_order;
 };
 
 /* for Playback/Capture */
diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index b2bd45e87bc3..4af09b1a0995 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -2520,6 +2520,7 @@ struct snd_soc_dai *snd_soc_register_dai(struct snd_soc_component *component,
 {
 	struct device *dev = component->dev;
 	struct snd_soc_dai *dai;
+	struct snd_soc_dai_ops *ops; /* REMOVE ME */
 
 	lockdep_assert_held(&client_mutex);
 
@@ -2548,6 +2549,30 @@ struct snd_soc_dai *snd_soc_register_dai(struct snd_soc_component *component,
 	if (!dai->name)
 		return NULL;
 
+	/* REMOVE ME */
+	if (dai_drv->probe		||
+	    dai_drv->remove		||
+	    dai_drv->compress_new	||
+	    dai_drv->pcm_new		||
+	    dai_drv->probe_order	||
+	    dai_drv->remove_order) {
+
+		ops = devm_kzalloc(dev, sizeof(struct snd_soc_dai_ops), GFP_KERNEL);
+		if (!ops)
+			return NULL;
+		if (dai_drv->ops)
+			memcpy(ops, dai_drv->ops, sizeof(struct snd_soc_dai_ops));
+
+		ops->probe		= dai_drv->probe;
+		ops->remove		= dai_drv->remove;
+		ops->compress_new	= dai_drv->compress_new;
+		ops->pcm_new		= dai_drv->pcm_new;
+		ops->probe_order	= dai_drv->probe_order;
+		ops->remove_order	= dai_drv->remove_order;
+
+		dai_drv->ops = ops;
+	}
+
 	dai->component = component;
 	dai->dev = dev;
 	dai->driver = dai_drv;
diff --git a/sound/soc/sunxi/sun50iw9-codec.c b/sound/soc/sunxi/sun50iw9-codec.c
index e9c4df24f0e0..38b1d3824c20 100644
--- a/sound/soc/sunxi/sun50iw9-codec.c
+++ b/sound/soc/sunxi/sun50iw9-codec.c
@@ -652,12 +652,9 @@ static int sun50i_h616_codec_dai_probe(struct snd_soc_dai *dai)
     return 0;
 }
 
-static const struct snd_soc_dai_ops dummy_dai_ops = {
-	.probe = sun50i_h616_codec_dai_probe,
-};
-
 static struct snd_soc_dai_driver dummy_cpu_dai = {
     .name = "sun50i_h616-codec-cpu-dai",
+    .probe = sun50i_h616_codec_dai_probe,
     .playback = {
         .stream_name = "Playback",
         .channels_min = 1,
diff --git a/sound/soc/sunxi_v2/snd_sunxi_ahub.c b/sound/soc/sunxi_v2/snd_sunxi_ahub.c
index e84c8b93fbcf..b53efd2c6d07 100644
--- a/sound/soc/sunxi_v2/snd_sunxi_ahub.c
+++ b/sound/soc/sunxi_v2/snd_sunxi_ahub.c
@@ -849,6 +849,21 @@ static void sunxi_ahub_dai_shutdown(struct snd_pcm_substream *substream,
 	}
 }
 
+static const struct snd_soc_dai_ops sunxi_ahub_dai_ops = {
+	/* call by machine */
+	.set_pll	= sunxi_ahub_dai_set_pll,	// set pllclk
+	.set_sysclk	= sunxi_ahub_dai_set_sysclk,	// set mclk
+	.set_bclk_ratio	= sunxi_ahub_dai_set_bclk_ratio,// set bclk freq
+	.set_tdm_slot	= sunxi_ahub_dai_set_tdm_slot,	// set slot num and width
+	.set_fmt	= sunxi_ahub_dai_set_fmt,	// set tdm fmt
+	/* call by asoc */
+	.startup	= sunxi_ahub_dai_startup,
+	.hw_params	= sunxi_ahub_dai_hw_params,	// set hardware params
+	.prepare	= sunxi_ahub_dai_prepare,	// clean irq and fifo
+	.trigger	= sunxi_ahub_dai_trigger,	// set drq
+	.shutdown	= sunxi_ahub_dai_shutdown,
+};
+
 static void snd_soc_sunxi_ahub_init(struct sunxi_ahub_info *ahub_info)
 {
 	struct regmap *regmap = NULL;
@@ -990,25 +1005,10 @@ static int sunxi_ahub_dai_remove(struct snd_soc_dai *dai)
 	return 0;
 }
 
-static const struct snd_soc_dai_ops sunxi_ahub_dai_ops = {
-	/* call by machine */
-	.set_pll	= sunxi_ahub_dai_set_pll,	// set pllclk
-	.set_sysclk	= sunxi_ahub_dai_set_sysclk,	// set mclk
-	.set_bclk_ratio	= sunxi_ahub_dai_set_bclk_ratio,// set bclk freq
-	.set_tdm_slot	= sunxi_ahub_dai_set_tdm_slot,	// set slot num and width
-	.set_fmt	= sunxi_ahub_dai_set_fmt,	// set tdm fmt
-	/* call by asoc */
-	.probe		= sunxi_ahub_dai_probe,
-	.startup	= sunxi_ahub_dai_startup,
-	.hw_params	= sunxi_ahub_dai_hw_params,	// set hardware params
-	.prepare	= sunxi_ahub_dai_prepare,	// clean irq and fifo
-	.trigger	= sunxi_ahub_dai_trigger,	// set drq
-	.remove		= sunxi_ahub_dai_remove,
-	.shutdown	= sunxi_ahub_dai_shutdown,
-};
-
 static struct snd_soc_dai_driver sunxi_ahub_dai = {
 	.name		= "ahub_plat",
+	.probe		= sunxi_ahub_dai_probe,
+	.remove		= sunxi_ahub_dai_remove,
 	.playback = {
 		.stream_name	= "Playback",
 		.channels_min	= 1,
